### DISCUSSION TITLE
CC: Tweaked 1.101 for MC 1.16.x, 1.18.x and 1.19.x

### DISCUSSION THREAD

**Discussion by SquidDev** [cc-tweaked/CC-Tweaked#1198] [Announcements]
So funny story, back in March of this year my employer went bankrupt. This didn't, as one might expect, leave me out of a job - I was sort of in a quasi-employment; paid, just with nothing to do. And so, amidst searching for a new job, I fell off the wagon of my retirement and started working on some CC: Tweaked features again.

While most of my hacking away hasn't been very productive, there have been some [useful](https://github.com/cc-tweaked/CC-Tweaked/pull/1065) [things](https://github.com/cc-tweaked/CC-Tweaked/pull/1057) come out of it. This release contains the last of this work, which is just a general cleanup to remove _some_ of tech debt which has accumulated over the years.

That said, this release does have some useful features!

## `file_transfer` event
Back in 1.97.0 we added support for uploading files onto a computer by dragging and dropping them from your file browser. This wrote directly to the computer file system, bypassing CraftOS entirely.

We now instead queue a [`file_transfer`](https://tweaked.cc/event/file_transfer.html) event, which contains the list of files to upload.

The shell listens to this event, so if you're not running any other program, you can just drop the file and the UX should be largely the same.

![A screenshot of a ComputerCraft advanced computer. The user is uploading an image titled "PreviousLovelyTuberculosis.png".](https://user-images.githubusercontent.com/4346137/199347705-51eb916a-569a-4dcb-b245-1c8c18ba5edf.png)

There's also an `import` program which has the same functionality, if you're running a shell which doesn't support file transfers.

## No more hidden files!
Shell autocompletion has always been a bit irritating, as it decides to complete hidden files before anything else. @IvoLeal72 has added a `shell.autocomplete_hidden` setting (defaults to true) which only completes hidden files if you type a `.`!

---

As always, thanks to everyone who has contributed to this release, be that with code, feedback or bug reports!

**Comment by SquidDev** [comment]
Oh, one thing I forgot to mention: this will be the last feature release for 1.16 and 1.18. I'll still fix any major bugs for the next few months (and I dare say there will be some), but want to focus development time on 1.19 for now.

**Reply by Lupus590** [reply]
Should new feature PRs be made against the 1.19 branch or should we keep using 1.16?

**Reply by SquidDev** [reply]
Good point: 1.19 most likely. I've changed the default branch.

**Reply by MCJEModder2026** [reply]
please do some updates to 1.18.2, it is a "LTS modding version"
